src/lib/device-online-state.ts

```diff
@@ -0,0 +1,306 @@
+					);
+				}
+			})
+			.then((msg: any) => {
+				const { id, message } = msg;
+
+				return Bluebird.try(() => {
+					const { uuid, nextState } = JSON.parse(message);
+
+					// raise and event for the state change...
+					switch (nextState) {
+						case DeviceOnlineStates.Timeout:
+							this.scheduleChangeOfStateForDevice(
+								uuid,
+								DeviceOnlineStates.Timeout,
+								DeviceOnlineStates.Offline,
+								API_HEARTBEAT_STATE_TIMEOUT_SECONDS, // put the device into a timeout state if it misses it's scheduled heartbeat window... then mark as offline
+							);
+							this.deviceTimeout(uuid);
+							break;
+						case DeviceOnlineStates.Offline:
+							this.deviceOffline(uuid);
+							break;
+						default:
+							// nothing should be raised here...
+							break;
+					}
+				}).then(() =>
+					this.rsmq.deleteMessageAsync({
+						qname: DeviceOnlineStateManager.EXPIRED_QUEUE,
+						id: id,
+					}),
+				);
+			})
+			.catch((err: Error | NoMessagesToConsumeSignal) => {
+				if (err instanceof NoMessagesToConsumeSignal) {
+					return;
+				}
+
+				captureException(
+					err,
+					'DeviceStateManager ==> Error when consuming queue',
+				);
+			})
+			.finally(() => this.consume());
+	}
+
+	private scheduleChangeOfStateForDevice(
+		uuid: string,
+		currentState: DeviceOnlineStates,
+		nextState: DeviceOnlineStates,
+		delay: number, // in seconds
+	) {
+		// remove the old queued state...
+		return this.redis
+			.getAsync(`${DeviceOnlineStateManager.REDIS_NAMESPACE}:${uuid}`)
+			.then(value => {
+				const { id } = JSON.parse(value) as { id: string };
+
+				if (id) {
+					return this.rsmq
+						.deleteMessageAsync({
+							qname: DeviceOnlineStateManager.EXPIRED_QUEUE,
+							id,
+						})
+						.return();
+				}
+			})
+			.catch(noop) // ignore errors when deleting the old queued state and queue the new one...
+			.then(() => {
+				return this.rsmq.sendMessageAsync({
+					qname: DeviceOnlineStateManager.EXPIRED_QUEUE,
+					message: JSON.stringify({
+						uuid,
+						nextState,
+					}),
+					delay,
+				});
+			})
+			.then(id => {
+				return this.redis
+					.setAsync(
+						`${DeviceOnlineStateManager.REDIS_NAMESPACE}:${uuid}`,
+						JSON.stringify({
+							id,
+							currentState,
+						}),
+					)
+					.then(() =>
+						this.redis.expireAsync(
+							`${DeviceOnlineStateManager.REDIS_NAMESPACE}:${uuid}`,
+							delay + 5, // add a small 5 second buffer to allow some wiggle room on top of the timeout window...
+						),
+					);
+			});
+	}
+
+	public captureEventFor(uuid: string, timeoutSeconds: number) {
+		// see if we already have a queued state for this device...
+		return this.redis
+			.getAsync(`${DeviceOnlineStateManager.REDIS_NAMESPACE}:${uuid}`)
+			.then(value => {
+				const { id, currentState } = JSON.parse(value);
+
+				if (!id || DeviceOnlineStates.Online !== currentState) {
+					this.deviceOnline(uuid);
+				}
+			})
+			.catch(() => noop)
```

Comment:
Pretty sure this should be
```suggestion
			.catch(noop)
```